File: haskell/words/lib/Numerical.hs
module Numerical (fibonacci, factorial, gcdy, lcmy) where

fibonacci :: Int -> Int
fibonacci n = fibonacciCalc n 0 1
  where
    fibonacciCalc :: Int -> Int -> Int -> Int
    fibonacciCalc x acc2 acc1
      | x < 0 = -1
      | x <= 1 = n
      | x == 2 = acc1 + acc2
      | otherwise = fibonacciCalc (x - 1) acc1 (acc1 + acc2)

factorial :: Int -> Int
factorial n = factorialCalc n 1
  where
    factorialCalc :: Int -> Int -> Int
    factorialCalc x acc
      | x < 0 = -1
      | x <= 1 = acc
      | otherwise = factorialCalc (x - 1) (x * acc)

gcdy :: Int -> Int -> Int
gcdy a b =
  if b == 0 then
    a
  else
    gcdy b (mod a b)

lcmy :: Int -> Int -> Int
lcmy a b = (a `div` gcdy a b) * b

File: ruby/words/src/numerical.rb
module Numerical
  def self.fibonacci(n)
    return -1 if n < 0
    return n if n <= 1
    acc1, acc2 = 1, 0
    (3..n).each do |i|
      acc1, acc2 = acc1 + acc2, acc1
    end
    acc1 + acc2
  end

  def self.factorial(n)
    return -1 if n < 0
    result = 1
    (2..n).each do |i|
      result *= i
    end
    result
  end

  def self.gcd(a, b)
    while b != 0
      a, b = b, a % b
    end
    a
  end

  def self.lcm(a, b)
    (a / gcd(a, b)) * b
  end
end